gdk/commands/component/PublishCommand.py

```diff
@@ -29,6 +34,181 @@ def run(self):
             )
             raise
 
+    def _check_for_changes(self):
+        logging.info(f"Checking for changes in the component: {self.project_config.component_name}")
+
+        options = self.project_config.options
+        only_on_change = options.get("only_on_change", [])
+
+        if only_on_change:
+            latest_published_recipe = self._get_latest_published_recipe()
+            if not latest_published_recipe:
+                logging.info(f"No published recipe found for the component: {self.project_config.component_name}")
+                return True
+
+            if "RECIPE" in only_on_change:
+                logging.info("Checking for changes in the RECIPE.")
+                if self._diff_recipe(latest_published_recipe):
+                    return True
+
+            if "ARTIFACTS" in only_on_change:
+                logging.info("Checking for changes in the ARTIFACTS.")
+                if self._diff_artifacts(latest_published_recipe):
+                    return True
+
+            logging.info(f"No changes in the component: {self.project_config.component_name}")
+            return False
+
+        else:
+            logging.info(f"Publishing regardless of RECIPE or ARTIFACT diff: {self.project_config.component_name}")
```

Comment:
I think we should default to publishing regardless of recipe/artifact diff if any specific component version is specified in the gdk-config file. IMO, we only care about checking diffs if the version is set to NEXT_PATCH, as this will keep incrementing the patch version and creating new components. If someone specifies an actual component version such as 1.0.0 or 2.0.0 we should attempt the publish to that version regardless of differences.

I'm thinking in some scenario someone may be developing with NEXT_PATCH (say along a 1.0.0 major/minor version), then decide they want to publish the same unchanged component to 2.0.0 to move up a major version and test with other components looking for a version in that dependency range. Unless they remove the only_on_change part of their config, we wouldn't make the publish as it would still be the same component as the latest patch version in the 1.0.0 major/minor version which is still latest. It seems more useful to me for GDK to attempt this publish regardless, rather than prevent a publish to an explicitly defined version. Of course, since only_on_change is not default, they would have added that to their config file at some point themselves as well, so I'm open to hearing your thoughts on what the behavior should be when version isn't set to NEXT_PATCH.

Reply:
Thinking as bellow for scenarios:

**version is NEXT_PATCH and only_on_change IS NOT set**
 - Then continue to publish as per existing behavior.

**version is NEXT_PATCH and only_on_change IS set**
 - Then check for changes against the latest published version.
 - only publish if there are changes

**version is NOT NEXT_PATCH and only_on_change IS NOT set**
 - Then continue to publish as per existing behavior.

**version is NOT NEXT_PATCH and only_on_change IS set** (this part was not in the initial PR)
 - Then check for changes against the published version that is set by version.
 - If the version does not exists publish
 - If the version does exists only publish if there are changes (more efficient)